File: backend/app/security.py
"""Passwords (Argon2id), JWT sessions, role permissions (SRS §4.10), secret encryption."""
import base64
import hashlib
import hmac
import json
import os
from datetime import datetime, timedelta, timezone

import jwt
from argon2 import PasswordHasher
from argon2.exceptions import VerifyMismatchError

from .config import settings

_ph = PasswordHasher()

def hash_password(pw: str) -> str:
    return _ph.hash(pw)

def verify_password(pw: str, hashed: str) -> bool:
    try:
        return _ph.verify(hashed, pw)
    except VerifyMismatchError:
        return False
    except Exception:
        return False


def create_token(user_id: str, org_id: str, role: str) -> str:
    payload = {
        "sub": user_id, "org": org_id, "role": role,
        "exp": datetime.now(timezone.utc) + timedelta(hours=settings.TOKEN_TTL_HOURS),
    }
    return jwt.encode(payload, settings.SECRET_KEY, algorithm="HS256")

def decode_token(token: str) -> dict | None:
    try:
        return jwt.decode(token, settings.SECRET_KEY, algorithms=["HS256"])
    except jwt.PyJWTError:
        return None


# --- Permission matrix (SRS §4.10) ---
ROLES = ("admin", "qa_lead", "qa_engineer", "viewer")

PERMISSIONS: dict[str, set[str]] = {
    "manage_members":      {"admin"},
    "manage_projects":     {"admin", "qa_lead"},
    "manage_environments": {"admin", "qa_lead"},
    "upload_documents":    {"admin", "qa_lead", "qa_engineer"},
    "edit_requirements":   {"admin", "qa_lead", "qa_engineer"},
    "import_spec":         {"admin", "qa_lead", "qa_engineer"},
    "generate":            {"admin", "qa_lead", "qa_engineer"},
    "edit_test_case":      {"admin", "qa_lead", "qa_engineer"},
    "approve_reject":      {"admin", "qa_lead"},
    "trigger_run":         {"admin", "qa_lead", "qa_engineer"},
    "view":                {"admin", "qa_lead", "qa_engineer", "viewer"},
    "export":              {"admin", "qa_lead", "qa_engineer", "viewer"},
    "view_audit_log":      {"admin", "qa_lead"},
}

def has_permission(role: str, capability: str) -> bool:
    return role in PERMISSIONS.get(capability, set())


# --- Environment secret encryption (NFR-SEC-02, simplified envelope) ---
def _derive_key() -> bytes:
    return hashlib.sha256(("traceo-secrets:" + settings.SECRET_KEY).encode()).digest()

def encrypt_secret(data: dict) -> bytes:
    """AES-GCM when 'cryptography' is available; HMAC-sealed XOR-stream fallback keeps
    the MVP dependency-light. Key custody outside the DB per NFR-SEC-02."""
    raw = json.dumps(data).encode()
    key = _derive_key()
    try:
        from cryptography.hazmat.primitives.ciphers.aead import AESGCM
        nonce = os.urandom(12)
        return b"AGCM" + nonce + AESGCM(key).encrypt(nonce, raw, None)
    except Exception:
        nonce = os.urandom(16)
        stream = hashlib.pbkdf2_hmac("sha256", key, nonce, 1, dklen=len(raw))
        ct = bytes(a ^ b for a, b in zip(raw, stream))
        mac = hmac.new(key, nonce + ct, hashlib.sha256).digest()
        return b"XSTR" + nonce + mac + ct

def decrypt_secret(blob: bytes | None) -> dict:
    if not blob:
        return {}
    key = _derive_key()
    try:
        if blob[:4] == b"AGCM":
            from cryptography.hazmat.primitives.ciphers.aead import AESGCM
            return json.loads(AESGCM(key).decrypt(blob[4:16], bytes(blob[16:]), None))
        if blob[:4] == b"XSTR":
            nonce, mac, ct = blob[4:20], blob[20:52], blob[52:]
            if not hmac.compare_digest(mac, hmac.new(key, nonce + ct, hashlib.sha256).digest()):
                return {}
            stream = hashlib.pbkdf2_hmac("sha256", key, nonce, 1, dklen=len(ct))
            return json.loads(bytes(a ^ b for a, b in zip(ct, stream)))
    except Exception:
        return {}
    return {}


SECRET_MASK = "••••••••"

def redact(text: str, secrets: list[str]) -> str:
    """Redaction at capture point (NFR-SEC-03) — never store an unredacted copy."""
    for s in secrets:
        if s and len(s) > 3:
            text = text.replace(s, SECRET_MASK)
    return text

File: backend-go/internal/security/security.go
// Package security: Argon2id passwords, JWT sessions, permission matrix (SRS §4.10),
// AES-GCM secret box, redaction-at-capture (NFR-SEC-03).
package security

import (
	"crypto/aes"
	"crypto/cipher"
	"crypto/rand"
	"crypto/sha256"
	"encoding/json"
	"errors"
	"strings"
	"time"

	"github.com/alexedwards/argon2id"
	"github.com/golang-jwt/jwt/v5"

	"traceo/internal/config"
)

func HashPassword(pw string) (string, error) {
	return argon2id.CreateHash(pw, argon2id.DefaultParams)
}

func VerifyPassword(pw, hash string) bool {
	ok, err := argon2id.ComparePasswordAndHash(pw, hash)
	return err == nil && ok
}

type Claims struct {
	Sub  string `json:"sub"`
	Org  string `json:"org"`
	Role string `json:"role"`
	jwt.RegisteredClaims
}

func CreateToken(userID, orgID, role string) (string, error) {
	c := Claims{
		Sub: userID, Org: orgID, Role: role,
		RegisteredClaims: jwt.RegisteredClaims{
			ExpiresAt: jwt.NewNumericDate(time.Now().Add(time.Duration(config.C.TokenTTLH) * time.Hour)),
		},
	}
	return jwt.NewWithClaims(jwt.SigningMethodHS256, c).SignedString([]byte(config.C.SecretKey))
}

func DecodeToken(tok string) (*Claims, error) {
	var c Claims
	t, err := jwt.ParseWithClaims(tok, &c, func(t *jwt.Token) (any, error) {
		if _, ok := t.Method.(*jwt.SigningMethodHMAC); !ok {
			return nil, errors.New("unexpected signing method")
		}
		return []byte(config.C.SecretKey), nil
	})
	if err != nil || !t.Valid {
		return nil, errors.New("invalid token")
	}
	return &c, nil
}

// ---- Permission matrix (SRS §4.10) — identical to the Python backend ----

var permissions = map[string][]string{
	"manage_members":      {"admin"},
	"manage_projects":     {"admin", "qa_lead"},
	"manage_environments": {"admin", "qa_lead"},
	"upload_documents":    {"admin", "qa_lead", "qa_engineer"},
	"edit_requirements":   {"admin", "qa_lead", "qa_engineer"},
	"import_spec":         {"admin", "qa_lead", "qa_engineer"},
	"generate":            {"admin", "qa_lead", "qa_engineer"},
	"edit_test_case":      {"admin", "qa_lead", "qa_engineer"},
	"approve_reject":      {"admin", "qa_lead"},
	"trigger_run":         {"admin", "qa_lead", "qa_engineer"},
	"view":                {"admin", "qa_lead", "qa_engineer", "viewer"},
	"export":              {"admin", "qa_lead", "qa_engineer", "viewer"},
	"view_audit_log":      {"admin", "qa_lead"},
}

func Has(role, capability string) bool {
	for _, r := range permissions[capability] {
		if r == role {
			return true
		}
	}
	return false
}

// ---- Secret box (environment auth configs) — AES-256-GCM, key derived from SecretKey ----

func boxKey() []byte {
	k := sha256.Sum256([]byte("traceo-secrets:" + config.C.SecretKey))
	return k[:]
}

func Encrypt(data map[string]any) []byte {
	raw, _ := json.Marshal(data)
	block, _ := aes.NewCipher(boxKey())
	gcm, _ := cipher.NewGCM(block)
	nonce := make([]byte, gcm.NonceSize())
	_, _ = rand.Read(nonce)
	return append(append([]byte("AGCM"), nonce...), gcm.Seal(nil, nonce, raw, nil)...)
}

func Decrypt(blob []byte) map[string]any {
	out := map[string]any{}
	if len(blob) < 4+12 || string(blob[:4]) != "AGCM" {
		return out
	}
	block, err := aes.NewCipher(boxKey())
	if err != nil {
		return out
	}
	gcm, err := cipher.NewGCM(block)
	if err != nil {
		return out
	}
	ns := gcm.NonceSize()
	raw, err := gcm.Open(nil, blob[4:4+ns], blob[4+ns:], nil)
	if err != nil {
		return out
	}
	_ = json.Unmarshal(raw, &out)
	return out
}

const SecretMask = "••••••••"

// Redact replaces secret values in text at capture point — never store unredacted.
func Redact(text string, secrets []string) string {
	for _, s := range secrets {
		if len(s) > 3 {
			text = strings.ReplaceAll(text, s, SecretMask)
		}
	}
	return text
}
